### ISSUE TITLE
StacKOverflowError when calling switchTo().window() using remote webdriver

### ISSUE THREAD

**Issue by Larry19b** [SeleniumHQ/selenium#6081]
## Description -
Selenium Version:  3.7.1

I am automating a java swing application and making use of remote web driver as follows:

        DesiredCapabilities dc = new DesiredCapabilities();
        dc.setBrowserName("java");
        dc.setVersion(" ");
        dc.setJavascriptEnabled(true);
        dc.setPlatform(Platform.ANY);
        javaDriverTest = new RemoteWebDriver(new URL("docker container url"), dc);

Everything works fine apart from the following:
        javaDriverTest.switchTo().window(windowName);

## Expected Behavior -
Window is switched successfully

## Actual Behavior -
StackOverflowError is thrown as follows:
```
java.lang.StackOverflowError
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:171)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at org.apache.http.impl.io.SessionInputBufferImpl.streamRead(SessionInputBufferImpl.java:137)
	at org.apache.http.impl.io.SessionInputBufferImpl.fillBuffer(SessionInputBufferImpl.java:153)
	at org.apache.http.impl.BHttpConnectionBase.fillInputBuffer(BHttpConnectionBase.java:344)
	at org.apache.http.impl.BHttpConnectionBase.isStale(BHttpConnectionBase.java:364)
	at org.apache.http.impl.conn.CPoolProxy.isStale(CPoolProxy.java:114)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:208)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:71)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:55)
	at org.openqa.selenium.remote.internal.ApacheHttpClient.fallBackExecute(ApacheHttpClient.java:138)
	at org.openqa.selenium.remote.internal.ApacheHttpClient.execute(ApacheHttpClient.java:86)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:161)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:600)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:657)
	at org.openqa.selenium.remote.RemoteWebDriver.getWindowHandles(RemoteWebDriver.java:505)
	at org.openqa.selenium.remote.RemoteWebDriver$RemoteTargetLocator.window(RemoteWebDriver.java:962)
	at org.openqa.selenium.remote.RemoteWebDriver$RemoteTargetLocator.window(RemoteWebDriver.java:963)
	at org.openqa.selenium.remote.RemoteWebDriver$RemoteTargetLocator.window(RemoteWebDriver.java:963)
	at org.openqa.selenium.remote.RemoteWebDriver$RemoteTargetLocator.window(RemoteWebDriver.java:963)
	at org.openqa.selenium.remote.RemoteWebDriver$RemoteTargetLocator.window(RemoteWebDriver.java:963)
	at org.openqa.selenium.remote.RemoteWebDriver$RemoteTargetLocator.window(RemoteWebDriver.java:963)
	at org.openqa.selenium.remote.RemoteWebDriver$RemoteTargetLocator.window(RemoteWebDriver.java:963)
	at org.openqa.selenium.remote.RemoteWebDriver$RemoteTargetLocator.window(RemoteWebDriver.java:963)
	at org.openqa.selenium.remote.RemoteWebDriver$RemoteTargetLocator.window(RemoteWebDriver.java:963)
	at org.openqa.selenium.remote.RemoteWebDriver$RemoteTargetLocator.window(RemoteWebDriver.java:963)
	at org.openqa.selenium.remote.RemoteWebDriver$RemoteTargetLocator.window(RemoteWebDriver.java:963)
```

Note: The issue also occurred on Selenium 3.11.0

**Comment by barancev** [comment]
For issues please provide a [concise reproducible test case](http://sscce.org/) and describe what results you are seeing and what results you expect.

See [CONTRIBUTING.md](https://github.com/SeleniumHQ/selenium/blob/master/CONTRIBUTING.md#issue-contributions)

**Comment by Larry19b** [comment]
@barancev apologies for not mentioning this earlier. The issue is intermittent, so somewhat difficult to reproduce. However, it occurs when I call the switchTo().window() method using RemoteWebDriver i.e.

To Reproduce:
- Set up desired capabilities as described in the issue above
- Attempt to use the switchTo().window() call

Note: It may require several attempts to get the error

**Comment by barancev** [comment]
Please provide a test case that runs many times in a loop, if the failure is intermittent. What window do you switch from and to?

There must be an edge case. We run regression tests on each commit, including tests for window switching, and we never saw such an exception. You must be having something very special.

**Comment by Larry19b** [comment]
Update:
I've surrounding the failing code with a try catch as follows:

        int maxTries = 0;
        while(maxTries < 61){
            try {
                maxTries++;
                javaDriverTest.switchTo().window(windowName);
            }catch(StackOverflowError e) {
                System.out.println("Catch block reached for StackOverflowError");
                Thread.sleep(1000);
                if (maxTries == 60){
                    throw new StackOverflowError(e.getMessage());
                }
            }
        }

The switchTo call will be attempted for a maximum of 60 seconds. It seems to be working fine now.

**Comment by barancev** [comment]
A workaround is a good thing, but we really want to reproduce and then fix the issue. Unfortunately, I can't reproduce it. I tried hard to switch to the same window or between two windows in a loop, but got no failure after 1000 iterations. You must be having an interesting edge case.

Can you please show a stack trace obtained from the latest version (3.13 at the moment)?

**Comment by Larry19b** [comment]
@barancev apologies for the late reply. I had been spending time trying to look for a more permanent solution and also checking to see if it will be possible to try the test with selenium 3.13. The api for automating the java swing application which I am using can be found here: 
https://github.com/jalian-systems/marathonv5/releases/tag/v5.1.4.0

There seems to be some incompatibility issues between marathon and selenium 3.13. Furthermore, when I try to use a more recent version of marathon, the tests themselves do not work. It maybe that the creator of marathon may need to make it compatible.

However, back to original issue, it seems its related to the way I started up the application i.e. I initially used the following:
```
export DISPLAY=:1.0
javaws -J -Dmarathon.agent=file:/root/Downloads/marathon-java-agent-5.1.4.0.jar.........
```
but when I added Xvfb to the command i.e.
```
Xvfb :1.0
export DISPLAY=:1.0
javaws -J -Dmarathon.agent=file:/root/Downloads/marathon-java-agent-5.1.4.0.jar.........
```
it fixed the issue. Therefore, I believe this issue is indeed not related to RemoteWebDriver, but actually to way the application under test is started.
